gnu_dataset/maverik-6.5/doc/MFS/man3/MAV_HBB.3
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH MAV_HBB 3 "AIG" "29/Mar/102" "GNU Maverik v6.2"
.IX Title "MAV_HBB 3"
.UC
.IX Name "B<MAV_HBB> - SMS data structure."
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
\fBMAV_HBB\fR \- SMS data structure.
.SH "LEVEL"
.IX Header "LEVEL"
Maverik Level 2
.SH "SYNTAX"
.IX Header "SYNTAX"
.PP
.Vb 5
\& typedef struct {
\&  MAV_HBBCluster *root;
\&  int size;
\&  MAV_HBBPointer *pointer;
\&} MAV_HBB;
.Ve
.IX Xref "MAV_HBB" 

.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This data structure is supplied for information only.
 
This is the data structure used to implement the default \fIMAV_HBB\fR
SMS, namely, a hierarchy of objects (based on bounding boxes).
.PP
Users should not manipulate this data structure directly, but rather by
the execution of the relevant callback (for example, add object, remove
object) which act upon it. Similarly, users will rarely need to
explicitly define this data structure since the
\f(CW@mav_HBBNew\fR function return a pointer to a newly created and
initialized data structures. This routine act as the second argument to
the \fBmav_SMSNew\fR function. For example:
.PP
<cod> sms= \fImav_SMSNew\fR\|(mav_SMSClass_HBB, \fImav_HBBNew()\fR);
</cod>

.rn }` ''


gnu_dataset/maverik-6.5/src/SMS/mav_hbb.c
/*
   GNU Maverik - a system for managing display and interaction in 
              Virtual Environment applications.
   Copyright (C) 2008  Advanced Interfaces Group

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

   The authors can be contacted via:
   www   - http://aig.cs.man.ac.uk
   email - maverik@aig.cs.man.ac.uk
   mail  - Advanced Interfaces Group, Room 2.94, Kilburn Building, 
        University of Manchester, Manchester, M13 9PL, UK
*/


#include "mavlib_sms.h"
#include <stdlib.h>
#include <stdio.h>

MAV_SMSClass *mav_SMSClass_HBB;

int allow_interrupts= 0;
int mav_render_aborted;

MAV_HBB *mav_HBBNew (void) {
  MAV_HBB *new_hbb;

  new_hbb= mav_malloc (sizeof (MAV_HBB));
  new_hbb->root= NULL;
  new_hbb->size= 0;
  new_hbb->pointer= NULL;

  return new_hbb;
}

/* Routine to create an SMS instance of this class */

MAV_SMS *mav_SMSHBBNew(void)
{
  return mav_SMSNew(mav_SMSClass_HBB, mav_HBBNew());
}

void mav_setFrameInterrupts (int mode)
{
  if (mode)
    allow_interrupts= 1;
  else
    allow_interrupts= 0;
}

int mav_HBBPointerReset (MAV_SMS *sms) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);

  if (hbb->pointer == NULL)
	hbb->pointer= (MAV_HBBPointer *) mav_malloc (sizeof (MAV_HBBPointer));

  hbb->pointer->curr_cluster= hbb->root;

  return 1;
}

void mavlib_HBBUpLevel (MAV_HBBPointer *ptr) {
  MAV_HBBCluster *last_child;
  MAV_HBBChild *ch;

  if (ptr->curr_cluster->parent == NULL)
  /* completed traversal */
    ptr->curr_cluster= NULL;
  else {
    last_child= ptr->curr_cluster;
    ptr->curr_cluster= ptr->curr_cluster->parent;

  /* check for another child node */
    ch= ptr->curr_cluster->children;

    while (ch->child != last_child)
  /* if this fails something well wacky has gone on */
      ch= ch->next;

  /* move to next child */
    ch= ch->next;

    if (ch) ptr->curr_cluster= ch->child;
    else
  /* last child on this level so go up another level */
      mavlib_HBBUpLevel (ptr);
  }
}

void mavlib_HBBNextCluster (MAV_HBBPointer *ptr) {

  /* if current node has children go down 1 level */
  if (ptr->curr_cluster->children) {
    ptr->curr_cluster= ptr->curr_cluster->children->child;

  /* check there is an object at this level otherwise move to next cluster */
    if (ptr->curr_cluster->obj == NULL) mavlib_HBBNextCluster (ptr);
  } else {
  /* go back up one (or more) levels */
    mavlib_HBBUpLevel (ptr);

    if (ptr->curr_cluster)
      if (ptr->curr_cluster->obj == NULL) mavlib_HBBNextCluster (ptr);
  }
}

int mav_HBBObjectNext (MAV_SMS *sms, MAV_object **next_obj) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);

  if (hbb->pointer->curr_cluster == NULL)
    return 0;
  else {
    if (hbb->pointer->curr_cluster->obj == NULL) {
      mavlib_HBBNextCluster (hbb->pointer);
      if (hbb->pointer->curr_cluster == NULL)
	return 0;
    }

    *next_obj= hbb->pointer->curr_cluster->obj;
    mavlib_HBBNextCluster (hbb->pointer);
  }

  return 1;
}

void mavlib_HBBDeleteFunc (MAV_HBBCluster *node) {
  MAV_HBBChild *ch;
  MAV_HBBChild *next_ch;

  /* delete node and all objects it contains */

  if (node != NULL) {
    ch= node->children;

  /* delete children first */
    while (ch != NULL) {
    next_ch= ch->next;
    mavlib_HBBDeleteFunc (ch->child);
    ch= next_ch;
  }

  /* if there is an object at this level then delete it */
  if (node->obj != NULL)
    mav_objectDelete (node->obj);

  /* free HBB node */
  mav_free (node);
}
}

#if 0
void
mav_SMSHBBDelete (MAV_SMS *sms) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet(sms);

  mavlib_deletedSMS= sms;
  mavlib_HBBDeleteFunc (hbb->root);
  mav_free (sms);
  mavlib_deletedSMS= sms;
}

void
mavlib_HBBEmptyFunc (MAV_HBBCluster *node, MAV_SMS *sms) {
  MAV_HBBChild *ch;
  MAV_HBBChild *next_ch;

  /* delete node but keep its objects */

  if (node != NULL) {
    ch= node->children;

  /* delete children first */
    while (ch != NULL) {
      next_ch= ch->next;
      mavlib_HBBEmptyFunc (ch->child, sms);
      ch= next_ch;
    }

  /* make sure object tables are kept up to date */
    if ((node->obj != NULL)&&(mav_opt_objectTables))
      mavlib_objectTablesRemoveSMS (node->obj, sms);

  /* free node memory */
    mav_free (node);
  }
}

void
mav_HBBEmpty (MAV_SMS *sms) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);

  mavlib_HBBEmptyFunc (hbb->root, sms);
  hbb->root= NULL;
  hbb->size= 0;
  hbb->pointer= NULL;
}
#endif

int mavlib_HBBRemoveObjectFunc (MAV_HBBCluster *node, MAV_object *mo) 
{
  int found= 0;
  MAV_HBBCluster *current;
  MAV_HBBCluster *parent_node;
  MAV_HBBChild *ch;
  MAV_HBBChild *last_ch;
  MAV_HBBChild *next_ch;

  current= node;

    /* check current object */
    if (current->obj == mo) {
      /* found it */
      found= 1;
      /* delete it from this node */
      current->obj= NULL;

      /* delete from parent's children list */
      parent_node= current->parent;
      if (parent_node == NULL) {
  	  /* root node */
      } else {
	ch= parent_node->children;
	last_ch= NULL;
	while (ch->child != current) {
	  last_ch= ch;
	  ch= ch->next;
	}
	if (last_ch == NULL)
	  parent_node->children= ch->next;
	else
	  last_ch->next= ch->next;
	mav_free (ch);
	parent_node->num_children --;
      }

      /* update children to point to correct parent */
      ch= current->children;
      while (ch != NULL) {
	next_ch= ch->next;
	ch->next= parent_node->children;
	parent_node->children= ch;
	parent_node->num_children ++;
	ch->child->parent= parent_node;
	ch= next_ch;
      }
    } else {
      ch= current->children;
      while (!found && (ch != NULL)) {
	found= mavlib_HBBRemoveObjectFunc (ch->child, mo); ch= ch->next;
      }
    }

    return found;
}

int mav_HBBObjectRmv (MAV_SMS *sms, MAV_object *mo) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);
  int rv=0;

  if (hbb->size) {
    if (mavlib_HBBRemoveObjectFunc (hbb->root, mo)) {
      rv=1;
      hbb->size --;
      if (mav_opt_objectTables) mav_objectTablesSMSRmv (mo, sms);
    }
  }

  return rv;
}

int mavlib_HBBContainsObjectFunc (MAV_HBBCluster *node, MAV_object *mo) 
{
  MAV_HBBCluster *current;
  MAV_HBBChild *ch;
  int found= 0;

    /* start at root */
    current= node;

    if (current->obj == mo)  {
      /* found it */
      found= 1;
    } else {
      /* search this node's children */
      ch= current->children;
      while ((!found) && (ch != NULL)) {
	found= mavlib_HBBContainsObjectFunc (ch->child, mo);
	ch= ch->next;
      }
    }
    return (found);
}

int mav_HBBContainsObject (MAV_SMS *sms, MAV_object *mo) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);

  if (hbb->size)
    return mavlib_HBBContainsObjectFunc (hbb->root, mo);
  else
    return 0;
}

int mav_HBBSize (MAV_SMS *sms, int *sz) {
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);

  *sz= hbb->size;

  return 1;
}

void mavlib_HBBDoCluster (MAV_HBBCluster *node, MAV_drawInfo *di, MAV_SMSExecFn *fn) {
/* do complete cluster without culling */
  MAV_HBBChild *ch;

#if 0
  if (allow_interrupts && !mav_render_aborted)
    if (mav_eventWaiting ())
      mav_render_aborted= 1;
#endif

  if (node->obj != NULL) 
  /* do this node's object */
    (*(fn->fn)) (node->obj, di, fn->params);

  ch= node->children;
  while ((ch != NULL) && !mav_render_aborted) {
  /* draw children */
    mavlib_HBBDoCluster (ch->child, di, fn);
    ch= ch->next;
  }
}

void mavlib_HBBCullFunc (MAV_HBBCluster *node, int *corner_list, MAV_drawInfo *di, MAV_SMSExecFn *fn)
{
/* cull and draw cluster */
  MAV_HBBChild *ch;
  int test_result;

  /* check for interrupt */
#if 0
  if (allow_interrupts && !mav_render_aborted)
    if (mav_eventWaiting ());
      mav_render_aborted= 1;
#endif

  if (!mav_render_aborted) {
  /* check this node's bb against the frustum */
    test_result= mav_BBIntersectsClipPlanes(node->box, corner_list, &di->cp);

  /* set status (for redraw) */
    node->status= test_result;

    if (test_result) {
  /* definitely intersected */
      if (test_result==1)
  /* total enclosure so draw complete cluster */
        mavlib_HBBDoCluster (node, di, fn);
      else {
  /* partial enclosure so do this object and intersect */
  /* rest of cluster */
        if (node->obj != NULL)
	  (*(fn->fn)) (node->obj, di, fn->params);

  /* do children */
	ch= node->children;
	while (ch != NULL) {
	  mavlib_HBBCullFunc (ch->child, corner_list, di, fn);
	  ch= ch->next;
	}
      }
    }
  }
}

int mav_HBBExecFn (MAV_SMS *sms, MAV_drawInfo *di, MAV_SMSExecFn *fn) 
{
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);
  int corner_list[MAV_MAX_CLIP_PLANES];
  int n;

  mav_render_aborted= 0;

  if (hbb->size) {
    for (n= 0; n <di->cp.num; n++)
      corner_list[n]= mav_BBGetCorner (di->cp.planes[n].norm);

    mavlib_HBBCullFunc (hbb->root, corner_list, di, fn);
  }

  return 1;
}

int mavlib_HBBIntersectHitBB (MAV_window *w, MAV_HBBCluster *mc, MAV_line ln, MAV_objectIntersection *objint, MAV_object **obj)
{
/* intersect a BB with a ray that we know hits it */
  MAV_objectIntersection objint2;
  MAV_HBBChild *ch;
  int rv= MAV_FALSE;
  int test_rv;

  if (mc->obj != NULL) {
  /* check this node's object */
    if (mav_callbackIntersectExec(w, mc->obj, ln, &objint2)) {
      rv= MAV_TRUE;

  /* check if valid hit */
      if (objint2.pt1 < objint->pt1) {
	*objint= objint2;
	*obj= mc->obj;
      }
    }
  }

  /* check this node's children */
  ch= mc->children;
  while (ch != NULL) {
  /* intersect child's bbox */
    if (mav_BBIntersectsLine (ch->child->box, ln, &objint2))
      if (objint2.pt1 < objint->pt1) {
  /* if a valid hit then recurse */
        test_rv= mavlib_HBBIntersectHitBB (w, ch->child, ln, objint, obj);
	if (!rv) rv= test_rv;
      }
    ch= ch->next;
  }

  return rv;
}

int mav_HBBIntersect (MAV_SMS *sms, MAV_window *w, MAV_line *ln, MAV_objectIntersection *objint, MAV_object **obj)
{
  MAV_HBB *hbb= (MAV_HBB *) mav_SMSDataGet (sms);
  MAV_objectIntersection objint2;
  int rv= MAV_FALSE;

  *obj= NULL;
  objint->pt1= MAV_INFINITY;

  if (hbb->size)
  /* intersect this node's bb */
    if (mav_BBIntersectsLine (hbb->root->box, *ln, &objint2))
      if (objint2.pt1 < objint->pt1)
        rv= mavlib_HBBIntersectHitBB (w, hbb->root, *ln, objint, obj);

  return rv;
}



int mav_HBBPointerPush(MAV_SMS *s)
{
  if (mav_opt_output==MAV_VERBOSE) fprintf(stderr, "Warning: HBBPointerPush not implemented yet!\n");
  return 0;
}

int mav_HBBPointerPop(MAV_SMS *s)
{
  if (mav_opt_output==MAV_VERBOSE) fprintf(stderr, "Warning: HBBPointerPop not implemented yet!\n");
  return 0;
}

int mav_HBBEmpty(MAV_SMS *s, int *o)
{
  if (mav_opt_output==MAV_VERBOSE) fprintf(stderr, "Warning: HBBEmpty not implemented yet!\n");
  return 0;
}

int mav_HBBDelete(MAV_SMS *s, int *o)
{
  if (mav_opt_output==MAV_VERBOSE) fprintf(stderr, "Warning: HBBDelete not implemented yet!\n");
  return 0;
}

